make polymorphic variadic builtin functions heterogeneous

mw5h opened the pull request:
**tree: polymorphic variadic builtin functions are now heterogeneous**

Previously, builtin functions with the parameter type VariadicType{VarType: types.Any} required that all variable arguments have the same type. This is what Postgres documents as its behavior, but the reality is that Postgres polymorphic builtins accept any heterogeneous arguments. This patch changes our overload handling to allow heterogeneous arguments to CRDB builtins.

**Fixes:** https://github.com/cockroachdb/cockroach/issues/136295
**Release notes (bug fix):** PREPARE now accepts heterogeneous arguments to the concat, json_build_object, jsonb_build_object, json_build_array, jsonb_build_array functions.

**builtins: support polymorphic variadic arguments in concat_ws()**

Previously, concat_ws took a variable length argument list of strings. Postgres allows anyelement data types for concat_ws(). Rewrite concat_ws() so that it takes a fixed string first argument followed by a variable list of anyelement arguments.

**Release note (sql change):** The concat_ws builtin function will now accept non-string arguments in the second and later positions.

mw5h committed:
tree: polymorphic variadic builtin functions are now heterogeneous

Previously, builtin functions with the parameter type
VariadicType{VarType: types.Any} required that all variable arguments
have the same type. This is what Postgres documents as its behavior, but
the reality is that Postgres polymorphic builtins accept any
heterogeneous arguments. This patch changes our overload handling to
allow heterogeneous arguments to CRDB builtins.

Fixes: #136295
Release notes (bug fix): PREPARE now accepts heterogeneous arguments to
the concat, json_build_object, jsonb_build_object, json_build_array,
jsonb_build_array functions.

mw5h committed:
builtins: support polymorphic variadic arguments in concat_ws()

Previously, concat_ws took a variable length argument list of strings.
Postgres allows anyelement data types for concat_ws(). Rewrite
concat_ws() so that it takes a fixed string first argument followed by a
variable list of anyelement arguments.

Epic: none
Release note (sql change): The concat_ws builtin function will now
accept non-string arguments in the second and later positions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140643)
<!-- Reviewable:end -->

mw5h committed:
pgwirebase: treat anyelement data as string data

By making polymorphic system builtin functions heterogeneous, we open
ourselves to the possibility that a prepared statement will have untyped
placeholders. This change treats those placeholders as string data (how
very MySQL of us).

Fixes: #136295
Release note (sql change): Untyped data is now accepted over the wire
protocol.

michae2 reviewed (commented):
> Reviewed 7 of 7 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140643)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIXCcKWD9o18dUKXgbo:-OIXCcKWD9o18dUKXgbp:b-8902oh):*
after that slack conversation, I think we need to be careful about 'any' vs 'anyelement', and this should be 'any'
___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIXCtb69baC_BLF0Iah:-OIXCtb69baC_BLF0Iai:b-gmq43p) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9ac168a2cd5049261bbff80ba5800e30952db74/pkg/sql/pgwire/pgwirebase/encoding.go#L861)):*
> ```Go
> 	}
> 	switch id {
> 	case oid.T_text, oid.T_varchar, oid.T_unknown, oid.T_anyelement:
> ```

does PG have a different oid type for `any`?
___
*[`pkg/sql/sem/tree/overload.go` line 1208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIXC4T54j5JJsBRnvei:-OIXC4T54j5JJsBRnvej:b-v7jp06) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eb97eacbcafb40a62bf1cdba2b1f355d1a529f1/pkg/sql/sem/tree/overload.go#L1208)):*
> ```Go
> 	if numOverloads == 1 {
> 		routineType, _, _ := s.overloads[0].outParamInfo()
> 		if vt, ok := s.params[0].(VariadicType); ok && vt.VarType == types.Any && routineType == BuiltinRoutine {
> ```

as discussed in [slack](https://cockroachlabs.slack.com/archives/G01R3EYPQKA/p1738952662468689), instead of checking `routineType == BuiltinRoutine`, we might want to divide `types.Any` from a new `types.AnyElement`, since they seem to be [different pseudo-types in PG](https://www.postgresql.org/docs/current/datatype-pseudo.html)
___
*[`pkg/sql/sem/tree/overload.go` line 1429 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIWvvfoCVU-JVCFaNNl:-OIWvvfoCVU-JVCFaNNm:b-wic1z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eb97eacbcafb40a62bf1cdba2b1f355d1a529f1/pkg/sql/sem/tree/overload.go#L1429)):*
> ```Go
> 		// operators).
> 		for i, ok := s.placeholderIdxs.Next(0); ok; i, ok = s.placeholderIdxs.Next(i + 1) {
> 			if s.typedExprs[i] != nil && heterogeneousBuiltin {
> ```

Do we need this check? (since `homogeneousTyp` is set to nil when `heterogeneousBuiltin` is set to true)


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140643)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/pgwire/pgwirebase/encoding.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIXCtb69baC_BLF0Iah:-OIXdpL0AN377sJGXwM8:b2acwcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9ac168a2cd5049261bbff80ba5800e30952db74/pkg/sql/pgwire/pgwirebase/encoding.go#L861)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

does PG have a different oid type for `any`?
</blockquote></details>

Yeah, oid.T_any exists (in both code bases).
___
*[`pkg/sql/sem/tree/overload.go` line 1208 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIXC4T54j5JJsBRnvei:-OIXcZlC32-BLIyyU2GP:bclkpll) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eb97eacbcafb40a62bf1cdba2b1f355d1a529f1/pkg/sql/sem/tree/overload.go#L1208)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

as discussed in [slack](https://cockroachlabs.slack.com/archives/G01R3EYPQKA/p1738952662468689), instead of checking `routineType == BuiltinRoutine`, we might want to divide `types.Any` from a new `types.AnyElement`, since they seem to be [different pseudo-types in PG](https://www.postgresql.org/docs/current/datatype-pseudo.html)
</blockquote></details>

Yikes. So, you're suggesting that we change the current types.Any into types.AnyElement and then introduce a new types.Any that is only used for builtin functions?
___
*[`pkg/sql/sem/tree/overload.go` line 1429 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140643#-OIWvvfoCVU-JVCFaNNl:-OIXab2306Mtu4PnKIAT:b-fyez3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eb97eacbcafb40a62bf1cdba2b1f355d1a529f1/pkg/sql/sem/tree/overload.go#L1429)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we need this check? (since `homogeneousTyp` is set to nil when `heterogeneousBuiltin` is set to true)
</blockquote></details>

homgoneousTyp can be set again at line 1396. I frankly don't like that very much (I don't like to reassign variables if possible), but I'm a bit afraid to monkey too much with this code because the knock-on effects can be subtle.


<!-- Sent from Reviewable.io -->

michae2 commented:
(just so we don't forget, [here's the relevant slack conversation](https://cockroachlabs.slack.com/archives/G01R3EYPQKA/p1738952662468689))
